File: 백준/Silver/1991. 트리 순회/트리 순회.java
import java.util.*;
import java.io.*;
public class Main {
	
	static class Node{
		Node left;
		Node right;
		char value;
		Node( char value, Node left, Node right){
			this.value = value;
			this.left = left;
			this.right = right;
		}	
	}	
	
	static void insertNode(Node node,char parent, char left, char right) {
		
		//첫번째값, 부모인지 확인
		if(node.value == parent) {
			//부모이면 자식노드를 생성하면서 값을 넣어줌 
			if(left == '.') 
				node.left = null;
			else 
				node.left = new Node(left,null,null);
			if(right == '.')
				node.right = null;
			else
				node.right = new Node(right,null,null);
		}
		
		//
		else {
			if(node.left!=null) insertNode(node.left,parent,left,right);
			if(node.right!=null) insertNode(node.right,parent,left,right);
		}
			
	}
	
	static void preOrder(Node n) {
		if(n==null)	return;
		System.out.print(n.value);
		preOrder(n.left);
		preOrder(n.right);
		
	}
	static void inOrder(Node n) {
		if(n==null)	return;
		inOrder(n.left);
		System.out.print(n.value);
		inOrder(n.right);
		
	}
	static void postOrder(Node n) {
		if(n==null)	return;
		postOrder(n.left);
		postOrder(n.right);
		System.out.print(n.value);
		
	}
	
	public static void main(String[] args) throws IOException{
		
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int N = Integer.parseInt(br.readLine());	//노드의 개수 
		Node root = new Node('A',null,null);
		
		for(int i=0;i<N;i++) {
			String [] str = br.readLine().split(" ");
			char parent = str[0].charAt(0);
			char left = str[1].charAt(0);
			char right = str[2].charAt(0);
			insertNode(root,parent,left,right);
		}
		
		preOrder(root);
		System.out.println();
		inOrder(root);
		System.out.println();
		postOrder(root);
	
	}// main()

}

File: 백준/Silver/1991. 트리 순회/트리 순회.py
class Node:
    def __init__(self, value, left,right):
        self.value = value
        self.left = left
        self.right = right


# preorder 전위순회   root 왼 오
def preorder(node):
    print(node.value,end='')
    if node.left != None:
        preorder(tree[node.left])
    if node.right != None:
        preorder(tree[node.right])

# inorder 중위순회  왼 root 오
def inorder(node):
    if node.left != None:
        inorder(tree[node.left])
    print(node.value,end='')
    if node.right != None:
        inorder(tree[node.right])


# postorder 후위순회  왼 오 root
def postorder(node):
    if node.left != None:
        postorder(tree[node.left])
    if node.right != None:
        postorder(tree[node.right])
    print(node.value,end='')

n = int(input())
tree = {}
for i in range(n):
    inputs = list(input().split())
    if inputs[1] == '.' : inputs[1] = None
    if inputs[2] == '.' : inputs[2] = None

    tree[inputs[0]] = Node(inputs[0], inputs[1],inputs[2])
    root = tree['A']

preorder(root)
print()
inorder(root)
print()
postorder(root)